parallel-build/src/worker-utils/cache.ts

```diff
@@ -0,0 +1,151 @@
+
+        const setReadFileCache = (key: Path, fileName: string) => {
+            const newValue = withTrace("read-disk", () => originalReadFile(fileName));
+            readFileCache.set(key, newValue !== undefined ? newValue : false);
+            return newValue;
+        };
+        host.readFile = fileName => {
+            const key = toPath(fileName);
+            const value = readFileCache.get(key);
+            if (value !== undefined) return value !== false ? value : undefined; // could be .d.ts from output
+
+            return setReadFileCache(key, fileName);
+        };
+
+        // fileExists for any kind of extension
+        host.fileExists = fileName => {
+            const key = toPath(fileName);
+            const value = fileExistsCache.get(key);
+            if (value !== undefined) return value;
+            const newValue = withTrace("fileExists-disk", () => originalFileExists(fileName));
+            fileExistsCache.set(key, !!newValue);
+            return newValue;
+        };
+        if (originalWriteFile) {
+            host.writeFile = (fileName, data, ...rest) => {
+                if (!suppressDiskWrite(fileName)) {
+                    withTrace("write-disk", () => originalWriteFile(fileName, data, ...rest));
+                }
+
+                const key = toPath(fileName);
+                fileExistsCache.set(key, true);
+
+                const value = readFileCache.get(key);
+                readFileCache.set(key, data);
+
+                if (value !== undefined && value !== data) {
+                    sourceFileCache.forEach(map => map.delete(key));
+                }
+                else {
+                    sourceFileCache.forEach(map => {
+                        const sourceFile = map.get(key);
+                        if (sourceFile && sourceFile.text !== data) {
+                            map.delete(key);
+                        }
+                    });
+                }
+            };
+        }
+
+        // directoryExists
+        if (originalDirectoryExists) {
+            host.directoryExists = directory => {
+                const key = toPath(directory);
+                const value = directoryExistsCache.get(key);
+                if (value !== undefined) return value;
+                const newValue = originalDirectoryExists(directory);
+                directoryExistsCache.set(key, !!newValue);
+                return newValue;
+            };
+
+            if (originalCreateDirectory) {
+                (host as any).createDirectory = (directory: string) => {
+                    originalCreateDirectory(directory);
+                    const key = toPath(directory);
+                    directoryExistsCache.set(key, true);
+                };
+            }
+        }
+        function cleanAST(node: Node) {
+            (node as any).id = undefined;
+            ts.forEachChild(node, cleanAST, nodes => nodes.forEach(cleanAST));
+        }

```

Comment:
This looks like an interesting trick. Does setting `node.id` to `undefined` trigger some kind of cache invalidation, e.g. the import resolution cache?